lib/src/algorithms.dart

```diff
@@ -16,14 +19,24 @@ import 'utils.dart';
 /// (`CastError` on some SDK versions).
 ///
 /// Returns -1 if [value] is not in the list by default.
-int binarySearch<T>(List<T> sortedList, T value, {int Function(T, T) compare}) {
-  compare ??= defaultCompare<T>();
+int binarySearch<E>(List<E> sortedList, E value, {int Function(E, E) compare}) =>
+    binarySearchBy<E, E>(sortedList, _id, compare ?? defaultCompare<E>(), value);
+
+
+/// Returns a position of the [value] in [sortedList], if it is there.
+///
+/// If the list isn't sorted according to the [compare] function on the [keyOf]
+/// property of the elements, the result is unpredictable.
+///
+/// Returns -1 if [value] is not in the list by default.
+int binarySearchBy<E, K>(List<E> sortedList, K keyOf(E element), int Function(K, K) compare, E value) {
```

Comment:
I don't think it's a good idea to add new top level methods to this package. Anything new should be added as an extension only. If we want to have these as top level methods internally for some reason, we should be hiding them from our exports.

Reply:
I completely disagree with not adding top-level functions.

Top-level functions are still a perfectly good way to introduce general reusable functionality.

Extension methods are less generally useful than top-level functions, but more convenient and discoverable.
That's why I have made the distinction of adding convenient and discoverable APIs using extension methods, and more rough and verbose reusable APIs as top-level functions.

Reply:
> Not every function deserves looking like an instance-method.

When I last looked it seemed like every function in _this_ packages deserves looking like an instance-method. If we wont' be adding `list.binarySearchBy` extension at all, that does change the discussion a bit, but I think that would be a good extension method.

> It also avoids the "all functions after 2.6 are extension methods, all functions before are not" issue.

I don't think that is an issue, I haven't seen anyone make that mistake.

The main thing I want to push on is to have 1 (recommended) way to do things. I want the users of this package to avoid negotiating during _their_ code review about whether to use the extension method version or the top-level method version.

> The general function doesn't really _fit in_ as an instance/extension member, but it's there in case you really need the full power.

I don't understand this. Is this ability to be passed as a function (with a _super_ specific function signature) rather than be torn off a specific list the "full power"?

Reply:
There are multiple layers of API here. Both a general low-level API that people can use to build other libraries from, and which isn't necessarily optimized for ease-of-use, and a high-level convenience API that accesses the low-level API in various convenient manners. Actually, most of the former until now. It's a package of building blocks more than a package of solutions.

We could stop exposing the low-level API directly. That would mean not exporting the new quickSort or xBy methods from `algorithms` and only expose them as extension methods. It'll probably work. There will likely be someone asking us to expose it anyway, but if the full power is still accessible as an extension method, it's probably not an issue.

As for `binarySearch`, the reasons I originally didn't make it as an extension is that it doesn't actually work on all lists. It only works on *sorted* lists, and therefore I didn't want it to be an instance method on all lists. (For the same reason I think `sort` is badly designed as well, but that can be saved by extension methods, binary search cannot).
It's ... probably overly cautious of me, and I'll add the binary search as extension methods as well.

Reply:
> There will likely be someone asking us to expose it anyway, but if the full power is still accessible as an extension method, it's probably not an issue.

Waiting to expose something until someone asks is my preferred approach. As a team we've spent a non-trivial amount of time trying to remove stuff that has little or no usage.

Reply:
We're now not exposing the new top-level members.